File: scripts/build_textbook_map.py
import json
import re
import os

base = r'C:\Users\haro\.openclaw\workspace\fe-exam-study\src\data'

cat_files = {
    'hardware': 'fe-hardware.ts',
    'software': 'fe-software.ts',
    'database': 'fe-database.ts',
    'network': 'fe-network.ts',
    'security': 'fe-security.ts',
    'management': 'fe-management.ts',
    'dev-methods': 'fe-dev-methods.ts',
    'project-mgmt': 'fe-project-mgmt.ts',
    'service-mgmt': 'fe-service-mgmt.ts',
}

tb_files = {
    'hardware': 'textbook-hardware.json',
    'software': 'textbook-software.json',
    'database': 'textbook-database.json',
    'network': 'textbook-network.json',
    'security': 'textbook-security.json',
    'management': 'textbook-management.json',
    'dev-methods': 'textbook-dev-methods.json',
    'project-mgmt': 'textbook-project-mgmt.json',
    'service-mgmt': 'textbook-service-mgmt.json',
}

def parse_ts_array(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    eq_idx = content.index('=')
    arr_start = content.index('[', eq_idx)

    depth = 0
    arr_end = arr_start
    for i in range(arr_start, len(content)):
        if content[i] == '[':
            depth += 1
        elif content[i] == ']':
            depth -= 1
            if depth == 0:
                arr_end = i + 1
                break

    arr = content[arr_start:arr_end]

    # Convert TS syntax to JSON
    # Step 1: Replace single-quoted strings with double-quoted
    result = []
    i = 0
    in_string = False
    string_char = None
    while i < len(arr):
        c = arr[i]
        if not in_string:
            if c == "'" or c == '"':
                in_string = True
                string_char = c
                result.append('"')  # Always use double quotes in output
            else:
                result.append(c)
        else:
            if c == string_char:
                if i + 1 < len(arr) and arr[i + 1] == string_char:
                    # Escaped quote
                    result.append(c)
                    result.append(c)
                    i += 1
                else:
                    in_string = False
                    result.append('"')
            elif c == '"':
                result.append('\\"')
            else:
                result.append(c)
        i += 1

    json_str = ''.join(result)

    # Step 2: Add quotes around unquoted keys (word followed by colon)
    json_str = re.sub(r'\b(\w+)\s*:', r'"\1":', json_str)

    # Step 3: Remove trailing commas before } or ]
    json_str = re.sub(r',(\s*[}\]])', r'\1', json_str)

    try:
        data = json.loads(json_str)
        return data
    except json.JSONDecodeError as e:
        print(f"  Parse error at pos {e.pos}: {e.msg}")
        # Show context
        start = max(0, e.pos - 50)
        end = min(len(json_str), e.pos + 50)
        print(f"  Context: ...{json_str[start:end]}...")
        return []


def main():
    result = {}

    for cat, ts_file in cat_files.items():
        ts_path = os.path.join(base, ts_file)
        tb_path = os.path.join(base, tb_files[cat])

        print(f"Processing {cat}...")

        questions = parse_ts_array(ts_path)
        print(f"  Parsed {len(questions)} questions")

        with open(tb_path, 'r', encoding='utf-8') as f:
            topics = json.load(f)

        # Build mapping
        tq = {}
        for topic in topics:
            tq[topic['topicId']] = []

        for q in questions:
            text_parts = [q.get('question', ''), q.get('explanation', '')]
            for opt in q.get('options', []):
                text_parts.append(opt.get('text', ''))
            text = ' '.join(text_parts).lower()

            for topic in topics:
                kws = [kw for kw in topic['keywords'] if len(kw) >= 2]
                score = 0
                for kw in kws:
                    if kw.lower() in text:
                        score += 1
                title_words = [w for w in re.split(r'[\(\)\uff0f]', topic['title']) if len(w) >= 2]
                for tw in title_words:
                    if tw.lower() in text:
                        score += 2
                if score >= 2:
                    tq[topic['topicId']].append(q['id'])

        matched = sum(1 for v in tq.values() if v)
        print(f"  Matched {matched}/{len(topics)} topics")

        result[cat] = tq

    out_path = os.path.join(base, 'textbook-question-map.json')
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(result, f, indent=2, ensure_ascii=False)

    print(f"\nSaved to {out_path}")


if __name__ == '__main__':
    main()

File: scripts/build-textbook-map.ts
import fs from 'fs';
import path from 'path';

// Load all textbook data
const textbookDir = 'src/data';
const textbookFiles = {
  hardware: 'textbook-hardware.json',
  software: 'textbook-software.json',
  database: 'textbook-database.json',
  network: 'textbook-network.json',
  security: 'textbook-security.json',
  management: 'textbook-management.json',
  'dev-methods': 'textbook-dev-methods.json',
  'project-mgmt': 'textbook-project-mgmt.json',
  'service-mgmt': 'textbook-service-mgmt.json',
};

// Category to ts file mapping
const categoryTsFiles: Record<string, string> = {
  hardware: 'fe-hardware.ts',
  software: 'fe-software.ts',
  database: 'fe-database.ts',
  network: 'fe-network.ts',
  security: 'fe-security.ts',
  management: 'fe-management.ts',
  'dev-methods': 'fe-dev-methods.ts',
  'project-mgmt': 'fe-project-mgmt.ts',
  'service-mgmt': 'fe-service-mgmt.ts',
};

type Question = {
  id: number;
  category: string;
  question: string;
  explanation: string;
  options: { text: string; correct: boolean }[];
  topicIds?: string[];
};

type TextbookTopic = {
  topicId: string;
  title: string;
  summary: string;
  keywords: string[];
};

// Build keyword -> topicId map for each category
function buildKeywordMap(category: string): Map<string, string[]> {
  const filepath = path.join(textbookDir, textbookFiles[category]);
  const data = JSON.parse(fs.readFileSync(filepath, 'utf-8')) as TextbookTopic[];
  const map = new Map<string, string[]>();

  for (const topic of data) {
    // Add title words
    const titleWords = topic.title.split(/[／\(\)]/).filter(w => w.length > 1);
    for (const w of titleWords) {
      add(map, w.trim(), topic.topicId);
    }
    // Add keywords
    for (const kw of topic.keywords) {
      add(map, kw, topic.topicId);
      // Also add shortened versions
      if (kw.length > 3) {
        add(map, kw.substring(0, kw.length - 1), topic.topicId);
      }
    }
  }
  return map;
}

function add(map: Map<string, string[]>, key: string, topicId: string) {
  const k = key.toLowerCase();
  if (k.length < 2) return;
  if (!map.has(k)) map.set(k, []);
  const arr = map.get(k)!;
  if (!arr.includes(topicId)) arr.push(topicId);
}

// Match questions to topics
function matchQuestions(questions: Question[], keywordMap: Map<string, string[]>, category: string): void {
  let matched = 0;
  let total = questions.length;

  for (const q of questions) {
    if (q.category !== category) continue;
    const text = (q.question + ' ' + q.explanation + ' ' + q.options.map(o => o.text).join(' ')).toLowerCase();
    const matchedTopics = new Set<string>();

    // Check all keywords
    for (const [keyword, topicIds] of keywordMap) {
      if (text.includes(keyword)) {
        for (const tid of topicIds) {
          matchedTopics.add(tid);
        }
      }
    }

    if (matchedTopics.size > 0) {
      q.topicIds = Array.from(matchedTopics);
      matched++;
    }
  }

  console.log(`${category}: ${matched}/${total} questions matched`);
}

// Main
const allQuestionsPath = 'src/data/questions.ts';
const content = fs.readFileSync(allQuestionsPath, 'utf-8');

// Parse questions from the TS file
// It exports an array from multiple imports
const categoryData: Record<string, Question[]> = {};

for (const [cat, tsFile] of Object.entries(categoryTsFiles)) {
  const tsPath = path.join(textbookDir, tsFile);
  const tsContent = fs.readFileSync(tsPath, 'utf-8');
  const match = tsContent.match(/export const questions:\s*Question\[\]\s*=\s*(\[[\s\S]*?\n\]);/);
  if (match) {
    try {
      // Remove TypeScript-specific syntax
      let jsonStr = match[1];
      // We need to use eval-like approach since it's TS
      // Instead, read from the compiled questions
      const allQContent = fs.readFileSync(allQuestionsPath, 'utf-8');
      const allMatch = allQContent.match(/export const allQuestions: Question\[\] = \[(.*)\];/s);
      if (allMatch) {
        console.log(`Found ${allMatch[1].substring(0, 100)}...`);
      }
    } catch (e) {
      console.error(`Error parsing ${cat}: ${e}`);
    }
  }
}

// Since TS files are hard to parse, let's read allQuestions directly
// The questions.ts re-exports from category files
console.log('Reading questions.ts...');
const qsMatch = content.match(/import.*\n/g);
console.log(`Found ${qsMatch?.length} imports`);

// Actually, let's just use the data JSON files or read each TS file properly
// For now, output the mapping as a separate JSON that the app can load

// Build mapping files
const topicQuestions: Record<string, Record<string, number[]>> = {};

for (const [cat, textbookFile] of Object.entries(textbookFiles)) {
  const tbPath = path.join(textbookDir, textbookFile);
  const topics = JSON.parse(fs.readFileSync(tbPath, 'utf-8')) as TextbookTopic[];
  const keywordMap = buildKeywordMap(cat);

  // Read questions from each TS file
  const tsPath = path.join(textbookDir, categoryTsFiles[cat]);
  let tsContent = fs.readFileSync(tsPath, 'utf-8');

  // Extract the questions array - simple approach: find the array
  const arrayStart = tsContent.indexOf('[');
  const arrayMatch = tsContent.match(/questions:\s*Question\[\]\s*=\s*\[/);

  if (arrayMatch) {
    let startIdx = tsContent.indexOf('[', arrayMatch.index! + arrayMatch[0].length);
    // Find the end of the array
    let depth = 1;
    let endIdx = startIdx + 1;
    while (depth > 0 && endIdx < tsContent.length) {
      if (tsContent[endIdx] === '[') depth++;
      else if (tsContent[endIdx] === ']') depth--;
      endIdx++;
    }
    let arrayStr = tsContent.substring(startIdx, endIdx);

    // Try to parse by replacing TS syntax
    // Remove trailing commas, etc.
    arrayStr = arrayStr.replace(/,\s*}/g, '}').replace(/,\s*]/g, ']');

    try {
      const questions: Question[] = eval(arrayStr);
      console.log(`${cat}: parsed ${questions.length} questions`);

      const topicQuestionMap: Record<string, number[]> = {};
      for (const topic of topics) {
        topicQuestionMap[topic.topicId] = [];
      }

      for (const q of questions) {
        const text = (q.question + ' ' + q.explanation).toLowerCase();
        for (const topic of topics) {
          const topicText = (topic.title + ' ' + topic.keywords.join(' ')).toLowerCase();
          // Check if any keyword from topic appears in question
          const kws = topic.keywords.filter(kw => kw.length >= 2);
          let score = 0;
          for (const kw of kws) {
            if (text.includes(kw.toLowerCase())) score++;
          }
          // Also check title words
          const titleWords = topic.title.split(/[／\(\)]/).filter(w => w.length >= 2);
          for (const tw of titleWords) {
            if (text.includes(tw.toLowerCase())) score += 2;
          }
          if (score >= 2) {
            topicQuestionMap[topic.topicId].push(q.id);
          }
        }
      }

      let topicMatched = Object.values(topicQuestionMap).filter(v => v.length > 0).length;
      console.log(`${cat}: ${topicMatched}/${topics.length} topics have matching questions`);

      topicQuestions[cat] = topicQuestionMap;
    } catch (e) {
      console.error(`${cat}: Error parsing questions: ${e}`);
    }
  }
}

// Save the mapping
const outputPath = 'src/data/textbook-question-map.json';
fs.writeFileSync(outputPath, JSON.stringify(topicQuestions, null, 2));
console.log(`\nSaved mapping to ${outputPath}`);
